src/libraries/System.Private.CoreLib/src/System/SearchValues/IndexOfAnyAsciiSearcher.cs

```diff
@@ -81,14 +94,87 @@ internal static unsafe void ComputeAsciiState<T>(ReadOnlySpan<T> values, out Asc
-                int lowNibble = value & 0xF;
+            if (!IsVectorizationSupported || values.Length > 16)
+            {
+                return false;
+            }
 
-                bitmapLocal[(uint)lowNibble] |= (byte)(1 << highNibble);
+            if (Ssse3.IsSupported && maxInclusive > 127)
+            {
+                // We could support [1, 254] if we did the "& 0xF" before calling into Shuffle in IndexOfAnyLookupUniqueLowNibbleCore.
+                // We currently optimize for the common case of ASCII characters instead, saving an instruction there.
+                return false;
             }
 
-            state = new AsciiState(bitmapSpace, lookupLocal);
+            if (typeof(T) == typeof(char))
+            {
+                if (maxInclusive >= byte.MaxValue)
+                {
+                    // When packing UTF-16 characters into bytes, values may saturate to 255 (false positives), hence ">=" instead of ">".
+                    return false;
+                }
+
+                if ((Ssse3.IsSupported || PackedSimd.IsSupported) && minInclusive == 0)
+                {
+                    // When packing UTF-16 characters into bytes, values may saturate to 0 (false positives).
+                    // We could also support 0 if we flowed Ssse3AndWasmHandleZeroInNeedle through.
+                    return false;
+                }
+            }
+
+            // We assume there are no duplicates to simplify the logic (if there are any, they just won't use this searching approach).
+            int seenNibbles = 0;
+
+            foreach (T tValue in values)
+            {
+                int bit = 1 << (int.CreateChecked(tValue) & 0xF);
+
+                if ((seenNibbles & bit) != 0)
+                {
+                    // We already saw a value with the same low nibble.
+                    return false;
+                }
+
+                seenNibbles |= bit;
+            }
+
+            return true;
+        }
+
+        public static void ComputeUniqueLowNibbleState<T>(ReadOnlySpan<T> values, out UniqueLowNibbleState state)
+            where T : struct, IUnsignedNumber<T>
+        {
+            Debug.Assert(typeof(T) == typeof(byte) || typeof(T) == typeof(char));
+
+            Vector128<byte> valuesByLowNibble = default;
+            BitVector256 lookup = default;
+
+            foreach (T tValue in values)
+            {
+                byte value = byte.CreateTruncating(tValue);
+                lookup.Set(value);
+                valuesByLowNibble.SetElementUnsafe(value & 0xF, value);
+            }
+
+            if (valuesByLowNibble.GetElement(0) == 0 && !lookup.Contains(0))
+            {
+                // Avoid false positives for the zero character if no other character has a low nibble of zero.
+                // We can replace it with any other byte that has a non-zero low nibble.
+                valuesByLowNibble.SetElementUnsafe(0, (byte)1);
```

Comment:
I didn't fully grok this. Why don't we need to check if 1 is already being used?

Reply:
All vector elements start out as 0, and not all of them may be initialized.

We map every input character to an element based on its lower nibble.
```
 0, 16, 32 ... => valuesByLowNibble[0]
 1, 17, 33 ... => valuesByLowNibble[1]
15, 31, 47 ... => valuesByLowNibble[15] 
```

The search works by first picking a potential match based on the low nibble (`Shuffle`) and then confirming it (`Equals`).

This means that input characters with a given low nibble only care about the element of `valuesByLowNibble` for that nibble. Values like 1 or 2 don't care about what the value of `valuesByLowNibble[7]` is since they'll never be mapped to it.

This also means that it's okay for `valuesByLowNibble` to be left uninitialized at 0.
The `Equals` could only match for an input character 0, but those will always be mapped to `valuesByLowNibble[0]` by the shuffle instead.

The edge case is the 0th nibble since the character 0 could be a false positive there.
But it'll only be a false positive if we don't have the character 0 in our values.
That's the `valuesByLowNibble.GetElement(0) == 0 && !lookup.Contains(0)` check above.

To avoid false positives for 0, we can use the same trick of setting the element to some "unreachable" value.
We can use any value with a non-zero nibble, as the shuffle will map any inputs with those values to a different element. 1 is just an arbitrary choice.